### ISSUE TITLE
using CTE with let (has fewer columns than were specified in the column list)

### ISSUE THREAD

**Issue by hambersmith** [linq2db/linq2db#1817]
```c#
var accountsInfo =
	from a in db.Accounts_
	where a.Id.In(accountIds)
	select new
	{
		a.Id,
		a.Number,
		a.Name,
		a.Type,
		incomeBalance = (decimal)(SqlFunctions.AccountBalanceGet(a.Id, issueDate.AddDays(-1), false) * SqlFunctions.AccountBalanceTypeGet((short)a.Type)),
		outgoingBalance = (decimal)(SqlFunctions.AccountBalanceGet(a.Id, date, false) *
									SqlFunctions.AccountBalanceTypeGet((short)a.Type))
					};

var accountsCte = accountsInfo.AsCte("accountsInfo");

var accInfo = (from cte in accountsCte
			let max = db.Transactions_.Where(_ => _.DebitAccountId == cte.Id)
				.Select(_ => _.OperDate)
				.Union(db.Transactions_.Where(_ => _.CreditAccountId == cte.Id).Select(_ => _.OperDate))
				 .Max()
				 select new AccountInfoDebtCalculationReport
				{
					AccountId = cte.Id,
					AccountNumber = cte.Number,
					AccountType = cte.Type,
					AccountTypeString = cte.Type.GetDisplayValue(),
					AccountName = cte.Name,
					LastOperDate = max,
					IncomeBalance = cte.incomeBalance,
					OutgoingBalance = cte.outgoingBalance
				 }).ToArray();
```

```
Exception: System.Data.SqlClient.SqlException
**Message  : 'accountsInfo' has fewer columns than were specified in the column list.**
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption, Boolean shouldCacheForAlwaysEncrypted)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.ExecuteReader(CommandBehavior behavior)
   at LinqToDB.Data.DbCommandProcessor.DbCommandDefaultProcessor.ExecuteReader(DbCommand cmd, CommandBehavior commandBehavior) in C:\GIT\signum\ThirdParty\Linq2DB\Data\DbCommandProcessor\DbCommandDefaultProcessor.cs:line 19
   at LinqToDB.Data.DbCommandProcessor.DbCommandProcessorExtensions.ExecuteReaderExt(IDbCommand cmd, CommandBehavior commandBehavior) in C:\GIT\signum\ThirdParty\Linq2DB\Data\DbCommandProcessor\DbCommandProcessorExtensions.cs:line 34
   at LinqToDB.Data.DataConnection.ExecuteReader(CommandBehavior commandBehavior) in C:\GIT\signum\ThirdParty\Linq2DB\Data\DataConnection.cs:line 1281
```

```sql
DECLARE @p1 Date
SET     @p1 = '2019-03-14T00:00:00'
DECLARE @date0 Date
SET     @date0 = '2019-07-19T00:00:00'

WITH [accountsInfo]
(
	[cte_field0], 
	[Id], 
	[Number], 
	[Type], 
	[cte_field0],
	[cte_field1], 
	[cte_field2], 
	[Name]
)
AS
(
	SELECT
		[a].[Id],
		[a].[Number],
		[a].[Type],
		[a].[Name],
		accounts.AccountBalanceGet([a].[Id], @p1, 0),
		accounts.AccountBalanceTypeGet([a].[Type]),
		accounts.AccountBalanceGet([a].[Id], @date0, 0)
	FROM
		[accounts].[Accounts] [a] WITH (NOLOCK)
	WHERE
		[a].[Id] IN (3901675, 3300115, 3891247, 4165505)
)
SELECT
	[cte].[Id],
	[cte].[Number],
	[cte].[Type],
	[cte].[Name],
	(
		SELECT
			Max([t1].[OperDate])
		FROM
			(
				SELECT
					[_1].[OperDate]
				FROM
					[accounts].[Transactions] [_1] WITH (NOLOCK)
				WHERE
					[_1].[DebitAccountId] = [cte].[cte_field0]
				UNION
				SELECT
					[_2].[OperDate]
				FROM
					[accounts].[Transactions] [_2] WITH (NOLOCK)
				WHERE
					[_2].[CreditAccountId] = [cte].[cte_field0]
			) [t1]
	),
	[cte].[cte_field0],
	[cte].[cte_field1],
	[cte].[cte_field2]
FROM
	[accountsInfo] [cte]
```

The problem seems to be using **let.* *
He believes that in the CTE there are **8** fields, not **7**.
Superfluous field highlighted in bold

WITH [accountsInfo]
(
	[cte_field0], 
	[Id], 
	[Number], 
	[Type], 
	**[cte_field0],**
	[cte_field1], 
	[cte_field2], 
	[Name]
)

If i delete this (**[cte_field0]**), the query works in SQL.

### Environment details
linq2db version: *2.7.4*
Database Server: *MSSQL 2017*
Operating system: *Windows 7*
Framework version: *.NET Framework 4.6.2*

**Comment by sdanyliv** [comment]
@hambersmith, could you please post Data Model involved in this Linq query?

**Comment by hambersmith** [comment]
```c#
	public partial class Account : ICrudEntity<long>
	{
		#region Simple columns

		public long Id { get; set; } // bigint
		public string Number { get; set; } // varchar(20)
		public string Name { get; set; } // varchar(5000)
		public AccountType Type { get; set; } // smallint
		public AccountState State { get; set; } // smallint

		#endregion
	}

	public partial class Transaction : ICrudEntity<long>
	{
		#region Simple columns

		public long Id { get; set; } // bigint
		public long RequestId { get; set; } // bigint
		public Date OperDate { get; set; } // date
		public long DebitAccountId { get; set; } // bigint
		public long CreditAccountId { get; set; } // bigint

		#endregion
	}
```

**Comment by sdanyliv** [comment]
I'll check it later, as a workaround try to rewrite this query in the following way:
```cs
var accountsInfo =
	from a in db.Accounts_
	where a.Id.In(accountIds)
	select new
	{
		a.Id,
		a.Number,
		a.Name,
		a.Type,
		incomeBalance = (decimal)(SqlFunctions.AccountBalanceGet(a.Id, issueDate.AddDays(-1), false) * SqlFunctions.AccountBalanceTypeGet((short)a.Type)),
		outgoingBalance = (decimal)(SqlFunctions.AccountBalanceGet(a.Id, date, false) *
									SqlFunctions.AccountBalanceTypeGet((short)a.Type))
					};

var accountsCte = accountsInfo.AsCte("accountsInfo");

var debitMaxDates = 
    from cte in accountsCte
    from t in db.Transactions.InnerJoin(t => t.DebitAccountId == cte.Id)
    group t by cte.Id into g
    select new { Id = g.Key, MaxDate = g.Max(_ => _. OperDate) };

var creditMaxDates =
    from cte in accountsCte
    from t in db.Transactions.InnerJoin(t => t.CreditAccountId == cte.Id)
    group t by cte.Id into g
    select new { Id = g.Key, MaxDate = g.Max(_ => _. OperDate) };

var maxDates = from d in debitMaxDates.Concat(creditMaxDates)
    group d by.Id into g
    select new { Id = g.Key, MaxDate = g.Max(_ => _. MaxDate) };

var accInfo = (from cte in accountsCte
			from d in maxDates.LeftJoin(d => d.Id == cte.Id)
			select new AccountInfoDebtCalculationReport
			{
				AccountId = cte.Id,
				AccountNumber = cte.Number,
				AccountType = cte.Type,
				AccountTypeString = cte.Type.GetDisplayValue(),
				AccountName = cte.Name,
				LastOperDate = Sql.ToNullable(d.MaxDate),
				IncomeBalance = cte.incomeBalance,
				OutgoingBalance = cte.outgoingBalance
			}).ToArray();
```
Unfortunately for this case CTE is not needed here :)

**Comment by sdanyliv** [comment]
This is my assumptions how such query should be built for performance purposes. But without execution plan and comparison - it is just my opinion.
There are several ways how to do that in SQL and it needs just analyzing execution plans.

**Comment by hambersmith** [comment]
CTE is faster (Reads 96) vs Reads 110 (sdanyliv variant)
![zzzzzzzzzzz](https://user-images.githubusercontent.com/48532909/61548837-464f7580-aa57-11e9-921b-0632b57dfb5a.png)

**Comment by sdanyliv** [comment]
So, looks like SQL Server optimizer much smarter than me :)
Interesting how big dataset do you have?

**Comment by hambersmith** [comment]
I tested on a small sample, about 10 accounts (accountIds) of 100 million tables

**Comment by hambersmith** [comment]
sdanyliv, any news about my problem with let in cte?

**Comment by sdanyliv** [comment]
Sorry, not yet. Will check today.

**Comment by sdanyliv** [comment]
@hambersmith, fix prepared and will be included in 2.8.0. Release planned this Thursday.

**Comment by MaceWindu** [comment]
Closed by #1821
